drivers/gpio/gpio_pcf8574.c

```diff
@@ -0,0 +1,375 @@
+static void pcf8574_int_gpio_handler(const struct device *dev, struct gpio_callback *gpio_cb,
+				     uint32_t pins)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(pins);
+
+	struct pcf8574_drv_data *drv_data =
+		CONTAINER_OF(gpio_cb, struct pcf8574_drv_data, int_gpio_cb);
+
+	k_work_submit(&drv_data->work);
+}
+
+/**
+ * @brief This function reads a value from the connected device
+ *
+ * @param dev Pointer to the device structure of the driver instance.
+ * @param value Pointer to a variable where pin values will be stored.
+ *
+ * @retval 0 If successful.
+ * @retval Negative value for error code.
+ */
+static int pcf8574_port_get_raw(const struct device *dev, gpio_port_value_t *value)
+{
+	struct pcf8574_drv_data *drv_data = dev->data;
+	int rc;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	/**
+	 * Reading of the input port also clears the generated interrupt,
+	 * thus the configured callbacks must be fired also here if needed.
+	 */
+	rc = pcf8574_process_input(dev, value);
+
+	k_sem_give(&drv_data->lock);
+
+	return rc;
+}
+
+/**
+ * @brief This function realizes the write connection to the i2c device.
+ *
+ * @param dev A pointer to the device structure
+ * @param mask A mask of bits to set some bits to LOW or HIGH
+ * @param value The value which is written via i2c to the pfc8574's output pins
+ * @param toggle A way to toggle some bits with xor
+ *
+ * @retval 0 If successful.
+ * @retval Negative value for error code.
+ */
+static int pcf8574_port_set_raw(const struct device *dev, uint8_t mask, uint8_t value,
+				uint8_t toggle)
+{
+	const struct pcf8574_drv_cfg *drv_cfg = dev->config;
+	struct pcf8574_drv_data *drv_data = dev->data;
+	int rc = 0;
+	uint8_t tx_buf;
+
+	/** Make sure, kernel isn't busy*/
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	tx_buf = (drv_data->pins_cfg & ~mask);
+	tx_buf |= (value & mask);
+	tx_buf ^= toggle;
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+	rc = i2c_write_dt(&drv_cfg->i2c, &tx_buf, sizeof(tx_buf));
+	k_sem_give(&drv_data->lock);
+
+	if (rc == 0) {
+		drv_data->pins_cfg = tx_buf;
+	}
+
+	if (rc != 0) {
+		LOG_ERR("%s: failed to write output port: %d", dev->name, rc);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief This function fills a dummy because the pfc8574 has no pins to configure.
+ * You can use it to set some pins permanent to HIGH or LOW until reset. It uses the port_set_raw
+ * function to set the pins of pcf8574 directly.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param pin The bit in the io register which is set to high
+ * @param flags The direction 0 for input 1 for output
+ *
+ * @retval 0 If successful.
+ * @retval Negative value for error.
+ */
+static int pcf8574_pin_configure(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	struct pcf8574_drv_data *drv_data = dev->data;
+	int ret = 0;
+	if (flags == 1) {
```

Comment:
`flags` is a bitfield with multiple options;

- input/output
- active high/low
- pull up/down
- ...

I don't think this is correctly implemented here.